python/dgl/data/csv_dataset.py

```diff
@@ -0,0 +1,420 @@
+                    dst_data[type].data[key] = value
+            for type, data in node_dict[graph_id].items():
+                assign_data(type, data['data'], graph.nodes)
+            for (type), data in edge_dict[graph_id].items():
+                assign_data(type, data['data'], graph.edges)
+            graph_dict[graph_id] = graph
+        return graph_dict
+
+
+class DefaultDataParser:
+    """ Default data parser for DGLCSVDataset. It
+        1. ignores any columns which does not have a header.
+        2. tries to convert to list of numeric values if cell data is a str.
+        3. read data and infer data type directly, otherwise.
+    """
+
+    def __init__(self, separator='|'):
+        self.separator = separator
+
+    def __call__(self, df):
+        data = {}
+        for header in df:
+            if 'Unnamed' in header:
+                dgl_warning("Unamed column is found. Ignored...")
+                continue
+            dt = df[header].to_numpy().squeeze()
+            if isinstance(dt[0], str):
+                #probably consists of list of numeric values
+                dt = np.array([ast.literal_eval(row) for row in dt])
+            data[header] = dt
+        return data
+
+
+class DGLCSVDataset(DGLDataset):
+    """ This class aims to parse data from CSV files, construct DGLGraph
+        and behaves as a DGLDataset.
+
+    Parameters
+    ----------
+    data_path : str
+        Directory which contains 'meta.yaml' and CSV files
+    force_reload : bool
+        Whether to reload the dataset. Default: False
+    verbose: bool
+        Whether to print out progress information. Default: True.
+    data_parser : dict
+        A dictionary used for data parsing when loading data from CSV files.
+        The key is node/edge/graph type which specifies the header in CSV file
+        and the value is a callable object which is used to parse corresponding
+        column data. Especially, the graph level data parser is required
+        to be placed under key '_GRAPH_'. Default: None. If None, a default
+        data parser is applied which load data directly and tries to convert
+        list into array.
+
+    Attributes
+    ----------
+    graphs : :class:`dgl.DGLGraph`
+        Graphs of the dataset
+    data : dict
+        any available graph-level data such as graph-level feature, labels.
+
+    Examples
+    --------
+    Single heterograph dataset
+    >>> csv_dataset = data.DGLCSVDataset(test_dir)
+    >>> g = csv_dataset[0]
+
+    Multiple graphs dataset
+    >>> csv_dataset = data.DGLCSVDataset(test_dir)
+    >>> g, label = csv_dataset[0]
+    """
+    META_YAML_NAME = 'meta.yaml'
+
+    def __init__(self, data_path, force_reload=False, verbose=True, data_parser=None):
+        """
+        data_parser : dict
+            {'user' : NodeDataParser,
+             ('user', 'like', 'item') : EdgeDataParser,
+             '_GRAPH_' : GraphDataParser,
+            }
```

Comment:
I will separate them. Have three arguments `node_feat_parser`, `edge_feat_parser` and `graph_feat_parser`.

Reply:
as even separate them, they are still dictionaries, so I merged them into one dict. both are ok for me.

Reply:
accepted.